File: haskell/src/DataStructures/DynamicArray.hs
{-|
Module      : DataStructures.DynamicArray
Description : Dynamic array implementation using Data.Vector
Stability   : stable

A dynamic array wrapper around Data.Vector.
Time: O(1) for access, O(n) for modification (creates new vector)
Space: O(n)
-}
module DataStructures.DynamicArray
  ( DynamicArray
  , empty
  , isEmpty
  , push
  , pop
  , get
  , set
  , insert
  , removeAt
  , indexOf
  , contains
  , size
  , fromList
  , toList
  ) where

import qualified Data.Vector as V

-- | A dynamic array backed by Vector
newtype DynamicArray a = DynamicArray (V.Vector a)
  deriving (Show, Eq)

-- | Create an empty dynamic array
empty :: DynamicArray a
empty = DynamicArray V.empty

-- | Check if array is empty
isEmpty :: DynamicArray a -> Bool
isEmpty (DynamicArray v) = V.null v

-- | Add element to the end
push :: a -> DynamicArray a -> DynamicArray a
push x (DynamicArray v) = DynamicArray (V.snoc v x)

-- | Remove and return the last element
pop :: DynamicArray a -> Maybe (a, DynamicArray a)
pop (DynamicArray v)
  | V.null v  = Nothing
  | otherwise = Just (V.last v, DynamicArray (V.init v))

-- | Get element at index
get :: Int -> DynamicArray a -> Maybe a
get i (DynamicArray v)
  | i < 0 || i >= V.length v = Nothing
  | otherwise                = Just (v V.! i)

-- | Set element at index
set :: Int -> a -> DynamicArray a -> Maybe (DynamicArray a)
set i x (DynamicArray v)
  | i < 0 || i >= V.length v = Nothing
  | otherwise                = Just (DynamicArray (v V.// [(i, x)]))

-- | Insert element at index
insert :: Int -> a -> DynamicArray a -> Maybe (DynamicArray a)
insert i x (DynamicArray v)
  | i < 0 || i > V.length v = Nothing
  | otherwise = Just $ DynamicArray $
      V.take i v V.++ V.singleton x V.++ V.drop i v

-- | Remove element at index
removeAt :: Int -> DynamicArray a -> Maybe (a, DynamicArray a)
removeAt i (DynamicArray v)
  | i < 0 || i >= V.length v = Nothing
  | otherwise = Just (v V.! i, DynamicArray $ V.take i v V.++ V.drop (i + 1) v)

-- | Find index of first occurrence
indexOf :: Eq a => a -> DynamicArray a -> Maybe Int
indexOf x (DynamicArray v) = V.findIndex (== x) v

-- | Check if element exists
contains :: Eq a => a -> DynamicArray a -> Bool
contains x (DynamicArray v) = V.elem x v

-- | Get the number of elements
size :: DynamicArray a -> Int
size (DynamicArray v) = V.length v

-- | Create from list
fromList :: [a] -> DynamicArray a
fromList = DynamicArray . V.fromList

-- | Convert to list
toList :: DynamicArray a -> [a]
toList (DynamicArray v) = V.toList v

File: go/datastructures/dynamic_array.go
package datastructures

// DynamicArray implements a growable array with amortized O(1) append.
// Time: O(1) amortized push, O(n) insert/remove
// Space: O(n)
type DynamicArray[T any] struct {
	data     []T
	size     int
	capacity int
}

func NewDynamicArray[T any](initialCapacity ...int) *DynamicArray[T] {
	cap := 8
	if len(initialCapacity) > 0 && initialCapacity[0] > 0 {
		cap = initialCapacity[0]
	}
	return &DynamicArray[T]{
		data:     make([]T, cap),
		capacity: cap,
	}
}

func (a *DynamicArray[T]) Push(value T) {
	if a.size == a.capacity {
		a.resize(a.capacity * 2)
	}
	a.data[a.size] = value
	a.size++
}

func (a *DynamicArray[T]) Pop() (T, bool) {
	var zero T
	if a.size == 0 {
		return zero, false
	}
	a.size--
	value := a.data[a.size]
	a.data[a.size] = zero
	if a.size > 0 && a.size == a.capacity/4 {
		a.resize(a.capacity / 2)
	}
	return value, true
}

func (a *DynamicArray[T]) Get(index int) (T, bool) {
	var zero T
	if index < 0 || index >= a.size {
		return zero, false
	}
	return a.data[index], true
}

func (a *DynamicArray[T]) Set(index int, value T) bool {
	if index < 0 || index >= a.size {
		return false
	}
	a.data[index] = value
	return true
}

func (a *DynamicArray[T]) Insert(index int, value T) bool {
	if index < 0 || index > a.size {
		return false
	}
	if a.size == a.capacity {
		a.resize(a.capacity * 2)
	}
	for i := a.size; i > index; i-- {
		a.data[i] = a.data[i-1]
	}
	a.data[index] = value
	a.size++
	return true
}

func (a *DynamicArray[T]) Remove(index int) (T, bool) {
	var zero T
	if index < 0 || index >= a.size {
		return zero, false
	}
	value := a.data[index]
	for i := index; i < a.size-1; i++ {
		a.data[i] = a.data[i+1]
	}
	a.size--
	a.data[a.size] = zero
	if a.size > 0 && a.size == a.capacity/4 {
		a.resize(a.capacity / 2)
	}
	return value, true
}

func (a *DynamicArray[T]) IsEmpty() bool {
	return a.size == 0
}

func (a *DynamicArray[T]) Size() int {
	return a.size
}

func (a *DynamicArray[T]) Capacity() int {
	return a.capacity
}

func (a *DynamicArray[T]) ToSlice() []T {
	result := make([]T, a.size)
	copy(result, a.data[:a.size])
	return result
}

func (a *DynamicArray[T]) Clear() {
	var zero T
	for i := 0; i < a.size; i++ {
		a.data[i] = zero
	}
	a.size = 0
}

func (a *DynamicArray[T]) resize(newCapacity int) {
	newData := make([]T, newCapacity)
	copy(newData, a.data[:a.size])
	a.data = newData
	a.capacity = newCapacity
}
